File: src/types/proposal.ts
/**
 * Proposal system for tracked changes.
 * Changes are represented as proposals, not diffs.
 * Accepting or rejecting a change is a first-class decision.
 */

import { NodeId, AnyNode, NodeStatus } from "./node.js";

export type ProposalStatus = "open" | "accepted" | "rejected" | "withdrawn";

export type OperationType = "insert" | "delete" | "update" | "move" | "create" | "status-change";

export interface ProposalMetadata {
  /** When this proposal was created */
  createdAt: string;
  /** Who created this proposal */
  createdBy: string;
  /** When this proposal was last modified */
  modifiedAt: string;
  /** Who last modified this proposal */
  modifiedBy: string;
  /** Optional description of why this change is proposed */
  rationale?: string;
}

/**
 * Base interface for all operations in a proposal.
 */
export interface Operation {
  /** Unique identifier for this operation */
  id: string;
  /** Type of operation */
  type: OperationType;
  /** Order of operations (for applying changes) */
  order: number;
}

/**
 * Create a new node.
 */
export interface CreateOperation extends Operation {
  type: "create";
  /** The node to create */
  node: AnyNode;
}

/**
 * Update an existing node.
 */
export interface UpdateOperation extends Operation {
  type: "update";
  /** ID of the node to update */
  nodeId: NodeId;
  /** Field-level changes */
  changes: {
    content?: string;
    status?: NodeStatus;
    [key: string]: unknown;
  };
}

/**
 * Delete a node (mark as deleted, preserve for provenance).
 */
export interface DeleteOperation extends Operation {
  type: "delete";
  /** ID of the node to delete */
  nodeId: NodeId;
  /** Reason for deletion */
  reason?: string;
}

/**
 * Change the status of a node.
 */
export interface StatusChangeOperation extends Operation {
  type: "status-change";
  /** ID of the node */
  nodeId: NodeId;
  /** New status */
  newStatus: NodeStatus;
  /** Previous status */
  oldStatus: NodeStatus;
  /** Reason for status change */
  reason?: string;
}

/**
 * Insert text at a specific position.
 */
export interface InsertOperation extends Operation {
  type: "insert";
  /** Position to insert at */
  position: number;
  /** Text to insert */
  text: string;
  /** Optional reference to source node */
  sourceNodeId?: NodeId;
}

/**
 * Delete text at a specific range.
 */
export interface DeleteTextOperation extends Operation {
  type: "delete";
  /** Start position */
  start: number;
  /** End position */
  end: number;
  /** Optional reference to source node */
  sourceNodeId?: NodeId;
}

/**
 * Move a node (change its position or parent).
 */
export interface MoveOperation extends Operation {
  type: "move";
  /** ID of the node to move */
  nodeId: NodeId;
  /** New position or parent */
  target: {
    position?: number;
    parentId?: NodeId;
  };
}

/**
 * Union type of all operations.
 */
export type AnyOperation =
  | CreateOperation
  | UpdateOperation
  | DeleteOperation
  | StatusChangeOperation
  | InsertOperation
  | DeleteTextOperation
  | MoveOperation;

/**
 * A proposal contains one or more operations that represent a change.
 */
export interface Proposal {
  /** Unique identifier for this proposal */
  id: string;
  /** Current status */
  status: ProposalStatus;
  /** Operations in this proposal */
  operations: AnyOperation[];
  /** Metadata */
  metadata: ProposalMetadata;
  /** Optional comments attached to this proposal */
  comments?: Comment[];
  /** Related proposal IDs (e.g., supersedes, conflicts with) */
  relations?: string[];
}

/**
 * A comment attached to a proposal or a specific operation.
 */
export interface Comment {
  /** Unique identifier */
  id: string;
  /** The comment text */
  content: string;
  /** Who made the comment */
  author: string;
  /** When the comment was made */
  createdAt: string;
  /** Optional reference to a specific operation */
  operationId?: string;
  /** Optional text range this comment refers to */
  textRange?: {
    start: number;
    end: number;
  };
  /** Replies to this comment */
  replies?: Comment[];
}

/**
 * Review action on a proposal.
 */
export interface Review {
  /** Unique identifier */
  id: string;
  /** The proposal being reviewed */
  proposalId: string;
  /** Who reviewed */
  reviewer: string;
  /** When reviewed */
  reviewedAt: string;
  /** Action taken */
  action: "accept" | "reject" | "request-changes";
  /** Optional comment */
  comment?: string;
  /** Which operations were accepted/rejected (if partial) */
  operationIds?: string[];
}

File: server/src/types/proposal.rs
//! Proposal, review, comment types.
//! Mirrors src/types/proposal.ts.

use serde::{Deserialize, Serialize};

use crate::types::{ContextNode, NodeId, NodeStatus};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ProposalStatus {
    Open,
    Accepted,
    Rejected,
    Withdrawn,
    /// Terminal: proposal has been applied to accepted truth. Prevents double-apply.
    Applied,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProposalMetadata {
    pub created_at: String,
    pub created_by: String,
    pub modified_at: String,
    pub modified_by: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rationale: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub required_approvers: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub approved_by: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub base_versions: Option<std::collections::HashMap<String, u32>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "kebab-case")]
pub enum Operation {
    Create {
        id: String,
        order: u32,
        node: ContextNode,
    },
    Update {
        id: String,
        order: u32,
        node_id: NodeId,
        changes: UpdateChanges,
    },
    Delete {
        id: String,
        order: u32,
        node_id: NodeId,
        #[serde(skip_serializing_if = "Option::is_none")]
        reason: Option<String>,
    },
    StatusChange {
        id: String,
        order: u32,
        node_id: NodeId,
        new_status: NodeStatus,
        old_status: NodeStatus,
        #[serde(skip_serializing_if = "Option::is_none")]
        reason: Option<String>,
    },
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct UpdateChanges {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub status: Option<NodeStatus>,
    #[serde(flatten)]
    pub extra: Option<std::collections::HashMap<String, serde_json::Value>>,
}

/// Metadata recorded when a proposal is applied. Required for audit and idempotency.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppliedMetadata {
    pub applied_at: String,
    pub applied_by: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub applied_from_review_id: Option<String>,
    pub applied_from_proposal_id: String,
    pub applied_to_revision_id: String,
    pub previous_revision_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Proposal {
    pub id: String,
    pub status: ProposalStatus,
    pub operations: Vec<Operation>,
    pub metadata: ProposalMetadata,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub comments: Option<Vec<Comment>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub relations: Option<Vec<String>>,
    /// Present only when status is Applied. Mandatory for audit and idempotent apply.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub applied: Option<AppliedMetadata>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CommentAnchor {
    pub node_id: NodeId,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub field: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub range: Option<CommentRange>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub quote: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CommentRange {
    pub start: u32,
    pub end: u32,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum CommentStatus {
    Open,
    Resolved,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Comment {
    pub id: String,
    pub content: String,
    pub author: String,
    pub created_at: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub status: Option<CommentStatus>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resolved_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resolved_by: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub operation_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub anchor: Option<CommentAnchor>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub replies: Option<Vec<Comment>>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ReviewAction {
    Accept,
    Reject,
    RequestChanges,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Review {
    pub id: String,
    pub proposal_id: String,
    pub reviewer: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reviewer_role: Option<String>,
    pub reviewed_at: String,
    pub action: ReviewAction,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub comment: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub comments: Option<Vec<Comment>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub operation_ids: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_approval: Option<bool>,
}
